### PR TITLE
Fix handling of search shard failures in the coordinating node

### PR DESCRIPTION
**Pull request by jimczi**
This commit ensures that we catch exceptions that happens when the coordinating
node consumes a shard response. Such failures can happen if we need to compute
the serialized size of an aggregation that was computed locally and the serialization
fails. In such case, we now return a shard error rather than an uncaught NPE for the entire
search.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/action/search/QueryPhaseResultConsumer.java [hunk 1]
```diff
@@ -140,7 +140,7 @@ public void consumeResult(SearchPhaseResult result, Runnable next) {
         }
         SearchPhaseController.ReducedQueryPhase reducePhase = controller.reducedQueryPhase(results.asList(), aggsList,
             topDocsList, topDocsStats, pendingMerges.numReducePhases, false, aggReduceContextBuilder, performFinalReduce);
-        if (hasAggs) {
+        if (hasAggs && reducePhase.aggregations != null) {
```
Comment (iverase) [inline server/src/main/java/org/elasticsearch/action/search/QueryPhaseResultConsumer.java:143]: Is it needed both conditions? I would expect that if reducePhase.aggregations is not null, then hasAggs is always true?

**Review by iverase — APPROVED** [review]
lgtm. I left a comment, maybe add a comment if both conditions are needed?

File: server/src/main/java/org/elasticsearch/action/search/QueryPhaseResultConsumer.java (hunk 1 above)
Comment (jimczi) [reply server/src/main/java/org/elasticsearch/action/search/QueryPhaseResultConsumer.java]: I added a comment in https://github.com/elastic/elasticsearch/pull/66280/commits/baabbbd38f395b7ca6adacd590a5475db647f168, it can be null if all shards failed.

File: server/src/main/java/org/elasticsearch/action/search/QueryPhaseResultConsumer.java [hunk 2]
```diff
@@ -290,9 +293,13 @@ long ramBytesUsedQueryResult(QuerySearchResult result) {
...
-            return result.aggregations()
-                .asSerialized(InternalAggregations::readFrom, namedWriteableRegistry)
-                .ramBytesUsed();
+            if (result.aggregations().isSerialized()) {
+                return result.aggregations()
+                    .asSerialized(InternalAggregations::readFrom, namedWriteableRegistry)
+                    .ramBytesUsed();
+            } else {
+                return result.aggregations().expand().getSerializedSize();
+            }
```
Comment (javanna) [inline server/src/main/java/org/elasticsearch/action/search/QueryPhaseResultConsumer.java:302]: I find this change hard to follow, I am sure I am missing something. Why do we need the conditional compared to before? And why do we call asSerialized when the aggregation is already serialized?

Comment (jimczi) [reply server/src/main/java/org/elasticsearch/action/search/QueryPhaseResultConsumer.java]: This is just to avoid serializing on a real buffer since we don't need to keep the result if the aggregations is already expanded. However I agree that it can be confusing so I pushed https://github.com/elastic/elasticsearch/pull/66280/commits/c8673076f2de921f4799de19de3e82d2c83e0149 to address your concerns. Does it look better ?

Comment (javanna) [reply server/src/main/java/org/elasticsearch/action/search/QueryPhaseResultConsumer.java]: yea much better thanks
